File: python/trie.py
from typing import List

class Node:
    def __init__(self):
        self.next: List[Node | None] = [None] * 26
        self.end = False

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, s: str):
        node = self.root
        for ch in s:
            ch = ord(ch) - ord('a')
            if not node.next[ch]:
                node.next[ch] = Node()
            node = node.next[ch]
        node.end = True
    
    def search(self, s:str):
        node = self.root
        for ch in s:
            ch = ord(ch) - ord('a')
            if not node.next[ch]:
                return False
            node = node.next[ch]
        return node.end

    def print(self):
        result = []
        self.dfs(self.root, "", result)
        print(result)

    def dfs(self, node, word, result):
        if node.end:
            result.append(word)
        for i in range(26):
            if node.next[i]:
                self.dfs(node.next[i], word+str(chr(i+ord('a'))), result)

File: rust/src/trie.rs
struct TrieNode {
    children: [Option<Box<TrieNode>>; 26],
    is_end: bool,
}

impl TrieNode {
    fn new() -> Self {
        Self {
            children: Default::default(),
            is_end: false,
        }
    }
}

pub struct Trie {
    root: TrieNode,
}

impl Trie {
    pub fn new() -> Self {
        Self { root: TrieNode::new() }
    }

    pub fn insert(&mut self, word: &str) {
        let mut node = &mut self.root;
        for b in word.bytes() {
            let idx = (b - b'a') as usize;
            node = node.children[idx]
                .get_or_insert_with(|| Box::new(TrieNode::new()));
        }
        node.is_end = true;
    }

    pub fn search(&self, word: &str) -> bool {
        let mut node = &self.root;
        for b in word.bytes() {
            let idx = (b - b'a') as usize;
            match &node.children[idx] {
                Some(next) => node = next,
                None => return false,
            }
        }
        node.is_end
    }
}
